### PR TITLE
ObjectBindingPatterns Implementation

### PR DESCRIPTION
**Pull request by hazzard993**
I've implemented Object Binding Patterns.

They can be used in methods, functions and declarations.

```ts
function test({x, y}) {
    print(x, y);
}
let {x, y} = {x: 0, y: 1};

test({x: 0, y: 1});     // Prints 0, 1
print(x, y);            // Prints 0, 1
```

They can have default values / initializers.

```ts
let {x = 5, y} = {y: 5};
print(x, y);            // Prints 5, 5
```

They can be nested.

```ts
let {x: {x, y}, z} = {x: {x: 0, y: 1}, z: 2};
print(x, y, z);         // Prints 0, 1, 2
```

They can have aliases.

```ts
let {x: foo, y: bar} = {x: 0, y: 1};
print(foo, bar);        // Prints 0, 1
```

They can also be exported.

```ts
export let {x, y} = {x: 0, y: 1};
```

I removed the throw from `transformVariableDeclaration` since TypeScript knows that all possible cases have been covered so it thinks the code is unreachable. It can be reached forcefully as evidenced in one specific test case. I removed this since TypeScript will never allow you to use a `false` keyword for a variable identifier.

### CODE REVIEW

File: src/LuaTransformer.ts [hunk 1]
```diff
@@ -1009,8 +1063,19 @@ export class LuaTransformer {
...
                 );
             }
-        } else {
-            throw TSTLErrors.UnsupportedKind("variable declaration", statement.name.kind, statement);
+        } else if (ts.isObjectBindingPattern(statement.name)) {
+            const statements = [];
+            let table: tstl.Identifier;
+            if (ts.isObjectLiteralExpression(statement.initializer)) {
+                // If destructuring a literal table, declare the table first
+                table = tstl.createIdentifier("_");
```
[the remark this answers is not in the document]
Comment (hazzard993) [reply src/LuaTransformer.ts]: I'll follow that convention, I'll fix that temporary variable up

File: src/LuaTransformer.ts [hunk 2]
```diff
@@ -766,6 +784,42 @@ export class LuaTransformer {
...
+                    tableExpression = tstl.createTableIndexExpression(tableExpression, propertyName);
+                });
+                const propertyNameString = tstl.createStringLiteral(propertyName.text);
+                let tableAccessExpression: tstl.Expression =
+                    tstl.createTableIndexExpression(tableExpression, propertyNameString);
+                if (element.initializer) {
+                    tableAccessExpression = tstl.createBinaryExpression(tableAccessExpression,
+                        this.transformExpression(element.initializer), tstl.SyntaxKind.OrOperator);
+                }
+                yield* this.createLocalOrExportedOrGlobalDeclaration(name, tableAccessExpression);
```
[the remark this answers is not in the document]
Comment (hazzard993) [reply src/LuaTransformer.ts]: This allows

```ts
export let {x, y} = {x: 0, y: 1};
```

to work. The function discovers the export flags in each identifier

File: src/LuaTransformer.ts [hunk 3]
```diff
@@ -737,6 +743,18 @@ export class LuaTransformer {
 
         headerStatements.push(...defaultValueDeclarations);
 
+        // Add object binding patterns
+        let identifierIndex = 0;
+        const objectBindingPatternDeclarations: tstl.Statement[] = [];
+        parameters.forEach(binding => {
+            if (ts.isObjectBindingPattern(binding.name)) {
```
[the remark this answers is not in the document]
Comment (hazzard993) [reply src/LuaTransformer.ts]: I can extend this pull request out to handle array binding patterns as well. I'll see if I can implement that too

File: src/LuaTransformer.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (tomblind) [reply src/LuaTransformer.ts]: oh, right - good call

**Comment by hazzard993** [comment]
Thanks for the reviews guys!

**Pull request merged**